### PR TITLE
Decentralizing: Removing big centralized ChannelKind

### PR DESCRIPTION
**Pull request by Yoric**
This PR entirely removes `ChannelKind`. This makes it much simpler to develop new Adapters out of our repo and removes one of the big blockers towards writing Adapters in languages other than Rust.

Cc @aosmond This will cause a number of breakage for REST clients.

r? @julienw for OpenZWave. Note that I have merged a few channels.
r? @mcav if you have time For Thinkerbell and the Thinkerbell adapter.
r? @dhylands for IP Camera
r? @cr For Phlipps Hue.
r? @fabricedesre For TTS and FoxBox

### CODE REVIEW

**Comment by aosmond** [comment]
r+ for WebPush, I like how it is taking shape :).

File: components/taxonomy/src/backend.rs [hunk 1]
```diff
@@ -696,6 +703,17 @@ impl State {
     /// registered, or a channel with the same identifier is already registered.
     /// In either cases, this method reverts all its changes.
     pub fn add_channel(&mut self, mut setter: Channel) -> Result<WatchRequest, Error> {
```
Comment (fabricedesre) [inline components/taxonomy/src/backend.rs]: uber nit: s/mut setter/mut channel

File: src/adapters/clock/mod.rs [hunk 2]
```diff
@@ -345,21 +345,39 @@ impl Clock {
         service.properties.insert("model".to_owned(), "Mozilla clock v1".to_owned());
         try!(adapt.add_service(service));
         try!(adapt.add_channel(Channel {
-            supports_watch: true,
-            supports_fetch: true,
-            kind: ChannelKind::CurrentTimeOfDay,
-            ..Channel::empty(&getter_time_of_day_id, &service_clock_id, &adapter_id)
+            implements: Id::new("clock/time-of-day-s"),
```
Comment (fabricedesre) [inline src/adapters/clock/mod.rs]: I would even prefer clock/time-of-day-seconds

**Comment by fabricedesre** [comment]
r=me for the tts adapter and the taxonomy router.

**Comment by mcav** [comment]
r=me for thinkerbell

**Comment by Yoric** [comment]
And an additional r? @azasypkin and @npark-mozilla for the integration tests.

File: components/openzwave-adapter/src/lib.rs [hunk 3]
```diff
@@ -304,9 +312,12 @@ impl OpenzwaveAdapter {
                         include_map.push(include_setter_id.clone(), controller);
 
                         box_manager.add_channel(Channel {
-                            kind: ChannelKind::ZwaveInclude,
-                            supports_send: true,
-                            ..Channel::empty(&include_setter_id, &service_id, &adapter_id)
+                            implements: TaxoId::new("zwave/include"),
```
Comment (azasypkin) [inline components/openzwave-adapter/src/lib.rs]: ugh, `implements` is reserved keyword in javascript, not a blocker, just kind of inconvenient sometimes.

Comment (Yoric) [inline components/openzwave-adapter/src/lib.rs]: Well, we can certainly rename it, if you have a better name in mind.

Comment (Yoric) [inline components/openzwave-adapter/src/lib.rs]: I guess I could rename it `feature`.

Comment (azasypkin) [inline components/openzwave-adapter/src/lib.rs]: Anything non-reserved (eg. `kind`, `type` or `feature`) will work well for client side, but don't bother if there is too much to rename :)

Maybe it's just me, but `channel's feature` makes me think for a moment that channel can have several `features` :)

**Comment by azasypkin** [comment]
r+ from the client app side, I've started to migrate client app to updated API in https://github.com/fxbox/app/pull/155, and everything works well so far.

**Comment by nojunpark** [comment]
r+ pending the travis pass after the conflict resolution:
it would be really nice to have this wiki updated too :)

https://wiki.mozilla.org/Connected_Devices/Projects/Project_Link/Taxonomy

**Comment by cr** [comment]
Sorry, still on PTO, so I'm a bit slow and just skimming the Hue-side changes. What I saw still looks good to me. r+

**Comment by Yoric** [comment]
Thanks everyone for the reviews!

File: components/openzwave-adapter/src/lib.rs [hunk 4]
```diff
@@ -304,9 +312,12 @@ impl OpenzwaveAdapter {
...
                         box_manager.add_channel(Channel {
-                            kind: ChannelKind::ZwaveInclude,
-                            supports_send: true,
-                            ..Channel::empty(&include_setter_id, &service_id, &adapter_id)
+                            feature: TaxoId::new("zwave/include"),
+                            supports_send: Some(Signature::accepts(Maybe::Required(Type::IsSecure))),
+                            id: include_setter_id.clone(),
+                            service: service_id.clone(),
+                            adapter: adapter_id.clone(),
+                            .. Channel::default()
```
Comment (julienw) [inline components/openzwave-adapter/src/lib.rs:320]: :+1:

File: components/openzwave-adapter/src/lib.rs [hunk 5]
```diff
@@ -340,36 +351,42 @@ impl OpenzwaveAdapter {
...
-                                ..Channel::empty(&getter_id, &node_id, &adapter_id)
-                            }).unwrap_or_else(|e| {
-                                error!("Couldn't add the getter {}: {}", value_id, e);
-                            });
+                        let chan = match kind {
+                            None => continue,
+                            Some(kind) => kind.clone()
+                        };
+
+                        let id = TaxoId::<Channel>::new(&value_id);
```
Comment (julienw) [inline components/openzwave-adapter/src/lib.rs:360]: nit: I don't think you need the `::<Channel>::`

File: components/openzwave-adapter/src/lib.rs [hunk 6]
```diff
@@ -340,36 +351,42 @@ impl OpenzwaveAdapter {
...
+
+                        let mut chan = Channel {
+                            id: id.clone(),
+                            service: node_id,
+                            adapter: adapter_id.clone(),
+                            ..chan
+                        };
+
+                        if vid.is_write_only() {
+                            // For some reason, the device is configured as not being readable.
```
Comment (julienw) [inline components/openzwave-adapter/src/lib.rs:370]: it's not the device, it's a specific value for a device.

File: components/openzwave-adapter/src/lib.rs [hunk 7]
```diff
@@ -340,36 +351,42 @@ impl OpenzwaveAdapter {
...
+                            // Make sure that the channel doesn't pretend the opposite.
+                            chan.supports_fetch = None;
+                            chan.supports_watch = None;
+                        } else {
+                            getter_map.push(id.clone(), vid);
+                        }
+                        if vid.is_read_only() {
+                            // For some reason, the device is configured as not being writeable.
```
Comment (julienw) [inline components/openzwave-adapter/src/lib.rs:378]: same here

**Comment by julienw** [comment]
Actuallly I hadn't review yet ;) I left only nits so it's good for me but take care next time. I know github doesn't make this easy...

**Comment by Yoric** [comment]
Oops, sorry.

I'm fixing this in https://github.com/fxbox/foxbox/pull/506/commits/70f1e0fe77e44f12bec6221877814ca0031b7baf

**Pull request merged**